frontend/simple/js/state.js

```diff
@@ -5,22 +5,123 @@
-  loggedIn: false
+  loggedInUser: null,
+  offset: [],
+  socket: null
 }
 
 // Mutations must be synchronous! Never call these directly!
 // http://vuex.vuejs.org/en/mutations.html
 const mutations = {
-  LOGIN (state) { state.loggedIn = true },
-  LOGOUT (state) { state.loggedIn = false }
+  login (state, user) { state.loggedInUser = user },
+  logout (state) { state.loggedInUser = null },
+  updateCurrentGroupLog (state, log) {
+    if (Array.isArray(log.offset)) {
+      state.currentGroupLog = log.log
+      state.offset = log.offset
+    } else {
+      state.currentGroupLog = log
+    }
+  },
+  updateAvailableGroups (state, available) {
+    state.availableGroups = available
+  },
+  updateSocket (state, socket) {
+    state.socket = socket
+  }
+}
+
+export const actions = {
+  async createGroup ({commit, state}, group) {
+    let log = await EventLog.createEventLogFromGroup(group)
+    let available = state.availableGroups
+    if (state.socket) {
+      await joinRoom(state.socket, group)
+    }
+    commit('updateCurrentGroupLog', log)
+    available.push(log.currentLogPosition)
+    commit('updateAvailableGroups', available)
+  },
+  // TODO: Connect this to user credentials
+  // limits to available groups for user
+  async loadGroup ({commit, state}, group) {
+    let available = state.availableGroups
+    if (available.find((grp) => { return group === grp })) {
+      if (state.socket) {
+        await joinRoom(state.socket, group)
+      }
+      let log = await EventLog.getEventLogForGroupId(group)
+      let offset = []
+      commit('updateCurrentGroupLog', {log, offset})
+    }
+  },
+  async appendLog ({commit, state}, event) {
+    let log = await EventLog.addItemToLog(state.currentGroupLog, event)
+    commit('updateCurrentGroupLog', log)
+  },
+  async receiveEvent ({commit, state}, msg) {
+    // check if this is a log update
+    if (msg.data && msg.data.entry && msg.data.entry.parentHash) {
+      let available = state.availableGroups
+      // check to see if this event is for an available group
+      if (available.find((grp) => { return msg.data.groupId === grp })) {
+        let log = await EventLog.getEventLogForGroupId(msg.data.groupId)
+        log = await EventLog.addItemToLog(log, msg.data.entry.data, true)
+        // update current if the event is for the current group
+        if (state.currentGroupLog && state.currentGroupLog.currentLogPosition === msg.data.groupId) {
+          commit('updateCurrentGroupLog', log)
+        }
+      }
+    }
+  },
+  async backward ({commit, state}) {
+    if (state.currentGroupLog) {
+      let entry = await EventLog.getItemFromLog(state.currentGroupLog, state.currentGroupLog.currentLogPosition)
+      let log = makeLog(state.currentGroupLog.groupId, entry.parentHash)
+      let offset = state.offset.slice(0)
+      offset.push(state.currentGroupLog.currentLogPosition)
+      commit('updateCurrentGroupLog', {log, offset})
```

Comment:
Am not 100% sure, but I don't think you need to clone the `offset` array, I think you can just do something like:

```js
commit('updateCurrentGroupLog', {log, offsetPush: true})
```

And then in the `updateCurrentGroupLog` mutation, without cloning the array, push to it.

Similarly, for the `forward` action below, you also don't need to create a copy (I don't think), and can just do:


```js
if (state.currentGroupLog && state.offset.length) {
  let prior = state.offset[state.offset.length - 1]
  let log = makeLog(state.currentGroupLog.groupId, prior)
  commit('updateCurrentGroupLog', {log, offsetPop: true})
}
```

And conduct the appropriate checks in `updateCurrentGroupLog`.